### DISCUSSION TITLE
Editor Quality of Life

### DISCUSSION THREAD

**Discussion by Methos-** [ppy/osu#18625] [Ideas] [answered]
Started using the editor in lazer since you made a hands down better timing tool, and it's hard to ignore using it. 

Enjoying the overall smoothness of manipulating patterns so far after placing them down, a few issues though.

1. When scaling objects, I understand you want us to use the onscreen selection box outlining the pattern, but there is currently, to my knowledge, no way to aspect lock this when scaling. If I make a perfect square and I scale it up, I want it to stay a square, not a rectangle, this goes for all patterns and this gets a bit confusing when working with angles. This also comes with the issue of, I want to make a pattern exactly x2 larger than the last, using the old client, I input a number and it gives me exactly that. Finding some way to reintroduce that would be beneficial I believe. Also helps when scaling the same pattern multiple times.

2. Another issue is we would use this distance snap number tool (shown below) to get distances down correctly. I understand it changed with BPM and slider velocity, but it still helped get distances uniform with eachother. I think bringing this back in some capacity one way or another would be appreciated.

![image](https://user-images.githubusercontent.com/6557053/172683668-aa7eadc6-0ee6-42bb-a5fe-63fb8abab519.png)

3. If you remove the timing attribute for a timing signature, in the timing menu, there is no way to reverse that decision, and you just have to reload the map, or of course do it again. I think it would be pretty intuitive to have this maybe baked into the undo hotkey. So I can just press CTRL-Z and restore my mistake.

4. There also seems to be no way to set a slider base speed for an entire map to work from, setting each slider's individual velocity is nice, but a bit tedious if all of them are 3.0x. I tried using maybe the scroll speed attribute but I don't think this is intended for osu! standard or just isn't working. Not quite sure.

**Comment by peppy** [comment, marked as answer]
Thanks for the feedback. Most of what you mention is definitely already something I'm aware of, but I'll answer with more detail as feasible.

> 1. When scaling objects, I understand you want us to use the onscreen selection box outlining the pattern, but there is currently, to my knowledge, no way to aspect lock this when scaling.

I'd definitely hope that aspect lock was default (with holding shift or something to override). I'd also expect there to be a way to enter a value. We'll likely do this in a way that can be reused in all places we need to be able to enter an arbitrary value for these kinds of modifications.

> 2. Another issue is we would use this distance snap number tool (shown below) to get distances down correctly. I understand it changed with BPM and slider velocity, but it still helped get distances uniform with eachother. I think bringing this back in some capacity one way or another would be appreciated.

I'd want the distance spacing to be shown inline, between objects or on a selection rectangle or something. Right now, when you click an object it *will* update the current velocity to that of the object, but as you mention it doesn't show the distance from last object (which is sometimes different).

> 3. If you remove the timing attribute for a timing signature, in the timing menu, there is no way to reverse that decision, and you just have to reload the map, or of course do it again. I think it would be pretty intuitive to have this maybe baked into the undo hotkey. So I can just press CTRL-Z and restore my mistake.

Undo/redo support is a bit flaky in the editor, due to how we are saving files (still using the legacy format). That said, there's no real excuse for it not working on the timing screen, and I too have been frustrated with this one while making a timing tutorial video, so expect it to be fixed as a priority.

> 4. There also seems to be no way to set a slider base speed for an entire map to work from, setting each slider's individual velocity is nice, but a bit tedious if all of them are 3.0x. I tried using maybe the scroll speed attribute but I don't think this is intended for osu! standard or just isn't working. Not quite sure.

Currently it will use the velocity of the previous object. We are also considering showing all recent snaps (see https://github.com/ppy/osu/discussions/18474) which may be relevant as well.
